### PR TITLE
Ticket4935 sans dls saving data

### PR DESCRIPTION
**Pull request by Alistair-McGann-Tessella**
Description of work
Data is saved to a file format which can be loaded in to the vendor's software for analysis.

Ticket 
https://github.com/ISISComputingGroup/IBEX/issues/4999

Acceptance criteria
- [ ] Data saved from the IOC can be loaded in to the vendor software
- [ ] The output file is saved in two places:
   - [ ] The data is saved where the archive can pick it up in the data directory.
   - [ ] The data is also saved in a user-specified location.

### CODE REVIEW

File: LSi_Correlator.py [hunk 1]
```diff
@@ -296,24 +328,41 @@ def take_data(self):
...
 
-        return "{filepath}/{filename}_{timestamp}".format(filepath=filepath, filename=filename, timestamp=timestamp)
+        if experiment_name == "":
+            filename = "{run_number}_{title}_{timestamp}.dat".format(
+                run_number=run_number, title=remove_non_ascii(g.get_title()), timestamp=timestamp
+            )
+        else:
+            filename = "{run_number}_{experiment_name}_{timestamp}.dat".format(
+                run_number=run_number, experiment_name=remove_non_ascii(experiment_name), timestamp=timestamp
+            )
```
Comment (John-Holt-Tessella) [inline LSi_Correlator.py]:
```suggestion
        if experiment_name == "":
            experiment_name = g.get_title()
        filename = "{run_number}_{experiment_name}_{timestamp}.dat".format(
                run_number=run_number, experiment_name=remove_non_ascii(experiment_name), timestamp=timestamp)
```

File: LSi_Correlator.py [hunk 2]
```diff
@@ -195,6 +225,8 @@ def update_pv_and_write_to_device(self, reason, value, update_setpoint: bool = F
                 if update_setpoint:
                     self.update_param_and_fields("{reason}:SP".format(reason=reason), new_pv_value)
 
+        self.updatePVs()
```
Comment (John-Holt-Tessella) [inline LSi_Correlator.py]: ? Why has this changed

File: PVConfig.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
+from collections import namedtuple
```
Comment (John-Holt-Tessella) [inline PVConfig.py]: Should this file have been deleted?

File: LSi_Correlator.py [hunk 4]
```diff
@@ -296,24 +328,41 @@ def take_data(self):
...
         """
+        filename = "{instrument}{run_number}_DLS_{timestamp}.txt"
+        timestamp = datetime.now().strftime("%Y-%m-%dT%H_%M_%S")
+        return filename.format(instrument=g.get_instrument(), run_number=g.get_runnumber(), timestamp=timestamp)
 
-        filepath = self.get_pv_value(Records.FILEPATH.name)
-        filename = self.get_pv_value(Records.FILENAME.name)
+    def get_user_filename(self):
+        """ Returns a default filename from the current run number and title """
+        experiment_name = self.get_pv_value(Records.EXPERIMENTNAME.name)
```
Comment (John-Holt-Tessella) [inline LSi_Correlator.py]: I don't like that we use get PV value to get a value from the LSI Correlator for this. Is there no other way to get this value?

File: LSi_Correlator.py [hunk 5]
```diff
@@ -322,37 +371,39 @@ def save_data(self, correlation, time_lags, trace_a, trace_b):
             time_lags (float array): The time lags
             trace_a (float array): The 'raw' trace A counts from the correlator
             trace_b (float array): The 'raw' trace B from the correlator
+            trace_time (float array): The elapsed time at which each raw data point was collected
         """
-
-        filename = self.add_timestamp_to_filename()
+        user_filename = os.path.join(self.user_filepath, self.get_user_filename())
+        archive_filename = os.path.join(DATA_DIR, self.get_archive_filename())
```
Comment (John-Holt-Tessella) [inline LSi_Correlator.py:377]: DATA DIR is never set what is it used for?

File: LSi_Correlator.py [hunk 6]
```diff
@@ -103,6 +124,15 @@ def __init__(self, host: str, firmware_revision: str, filepath: str):
 
         self.updatePVs()
 
+    def update_output_file(self):
+        """
+        Updates the PV containing the name of the user's data file
+        """
+        while True:
+            filename = self.get_user_filename()
+            self.update_pv_value_and_write_to_device(Records.OUTPUTFILE.name, filename)
```
Comment (John-Holt-Tessella) [inline LSi_Correlator.py]: Records.OUTPUTFILE I am not sure that this exists

File: LSi_Correlator.py [hunk 7]
```diff
@@ -296,24 +328,41 @@ def take_data(self):
                 self.set_array_pv_value(Records.CORRELATION_FUNCTION.name, corr)
                 self.set_array_pv_value(Records.LAGS.name, lags)
 
-                self.save_data(corr, lags, trace_a, trace_b)
+                # Time axis is number of data points collected * scaling factor
+                time_trace = np.arange(len(self.device.TraceChA))*DELTA_T
```
Comment (John-Holt-Tessella) [inline LSi_Correlator.py]:
```suggestion
                time_trace = np.arange(len(trace_a))*DELTA_T
```

**Comment by John-Holt-Tessella** [comment]
This also doesn't look right:

lags = lags[np.isfinite(corr)]

in LSI Correlator.py

File: LSi_Correlator.py (hunk 5 above)
Comment (Alistair-McGann-Tessella) [reply LSi_Correlator.py:377]: DATA_DIR is the location in which the data files for the archive are saved. It's a well-defined location, but I wasn't sure if the directory should be published in the code.

**Comment by Alistair-McGann-Tessella** [comment]
> 
> 
> This also doesn't look right:
> 
> lags = lags[np.isfinite(corr)]
> 
> in LSI Correlator.py

The time lags and correlation function are the X and Y dimensions of the same dataset. We are removing the NaN values in the correlation function, so we also need to remove the corresponding time lags.
Unless you're suggesting we just let the NaN values through? In which case we need to remove both lines

File: LSi_Correlator.py (hunk 4 above)
Comment (Alistair-McGann-Tessella) [reply LSi_Correlator.py]: I don't think there's any other way. EXPERIMENTNAME is an override the user can supply on the OPI to change the name of the user data file to whatever they want. If the user doesn't supply anything, the run title is used instead.

**Comment by John-Holt-Tessella** [comment]
> > This also doesn't look right:
> > lags = lags[np.isfinite(corr)]
> > in LSI Correlator.py
> 
> The time lags and correlation function are the X and Y dimensions of the same dataset. We are removing the NaN values in the correlation function, so we also need to remove the corresponding time lags.
> Unless you're suggesting we just let the NaN values through? In which case we need to remove both lines

Ok makes sense

File: LSi_Correlator.py [hunk 8]
```diff
@@ -27,6 +28,11 @@
 from server_common.utilities import print_and_log
 from server_common.ioc_data_source import IocDataSource
 from server_common.mysql_abstraction_layer import SQLAbstraction
+from server_common.channel_access import ChannelAccess
+from file_format import FILE_SCHEME
+from pathlib import Path
+
+DATA_DIR = ""
```
Comment (John-Holt-Tessella) [inline LSi_Correlator.py]: This should be set

File: LSi_Correlator.py (hunk 5 above)
Comment (John-Holt-Tessella) [reply LSi_Correlator.py:377]: That is fine, I am sure it is in many places already.

**Review by John-Holt-Tessella — APPROVED** [review]

**Pull request merged**